lib/spack/spack/fetch_strategy.py

```diff
@@ -907,9 +907,12 @@ def clone(self, dest=None, commit=None, branch=None, tag=None, bare=False):
         git_commands = []
         submodules = self.submodules
         if callable(submodules):
-            submodules = list(submodules(self.package))
-            git_commands.append(["submodule", "init", "--"] + submodules)
-            git_commands.append(["submodule", "update", "--recursive"])
+            submodules = submodules(self.package)
+            if submodules:
```

Comment:
Can you:
1. Extend the test added in #27293 to include this case
2. Update the type signature for the `submodules` arg of `version()` in directives.py?

I think (2) should be something like:

```python

SubmoduleCallback = Callable["spack.package_base.PackageBase", Union[str, List[str]]]

@directive("versions")
def version(
    ...,
    submodules: Union[SubmoduleCallback, bool] = False,
```